File: G/134-gas-station/main.go
package main

import "fmt"

func canCompleteCircuit(gas []int, cost []int) int {
	buf := make([]int, len(gas))
	sum := 0
	for i := 0; i < len(gas); i++ {
		buf[i] = gas[i] - cost[i]
		sum += buf[i]
	}
	if sum < 0 {
		return -1
	}
	first := -1
	for i := 0; i < len(buf); i++ {
		if buf[i] >= 0 {
			first = i
			break
		}
	}

	as := 0
	index := -1
	for i := first; i < first+len(buf); i++ {
		idx := i % len(buf)
		if buf[idx] >= 0 && index == -1 {
			as = 0
			index = idx
		}
		as += buf[idx]
		if as < 0 {
			index = -1
		}
	}
	return index
}

func main() {
	fmt.Println("a")
}

File: G/134-gas-station/main.rs
struct Solution {}

impl Solution {
  pub fn can_complete_circuit(gas: Vec<i32>, cost: Vec<i32>) -> i32 {
    let mut ans: i32 = -1;
    let mut sum: i32 = 0;
    let mut total: i32 = 0;
    (0..gas.len()).for_each(|idx| {
      let d = gas[idx] - cost[idx];
      sum += d;
      total += d;
      if sum < 0 {
        ans = -1;
        sum = 0;
      } else if ans == -1 {
        ans = idx as i32;
      }
    });
    match total < 0 {
      true => -1,
      _ => ans,
    }
  }
}
